include/pybind11/pybind11.h

```diff
@@ -56,6 +57,21 @@
 
 PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
 
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+#if defined(_MSC_VER)
```

Comment:
Instead of poping the warning, I think the recommended way to handle this error is actually to call the  ```_strdup``` on MSC: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strdup-wcsdup?view=msvc-160

Reply:
The 2nd thoughts I had when marking this PR as draft yesterday:

* Cluttering pybind11 with MSVC-specific things is ugly.
* Better to simply disable the warning on the command line?

To me it looks like just another MSVC oddity, dirt that I want to keep out of my house, to use stronger words.

But we have similar pros-and-cons for most of the 11 items that I enumerated here yesterday:
**[pybind11.h v2.7.0 pragma block analysis](https://docs.google.com/spreadsheets/d/1XDKkq1w7d9f2dPcVsn-vmZHleMWQkVCl7xb4cilpGiI/edit?usp=sharing)**

I was thinking of suggesting a focused action meeting to go through the list and decide what to do about each one. I'll ping you and @henryiii on slack about it.

Reply:
I think we should do what MSVC wants; like it or not, they are a major target for us. We should not requires any special flags to build without warnings; this is a header-only library and is not _always_ called through build methods we control.